test/plugins/test-trace-http.ts

```diff
@@ -13,357 +13,420 @@
+import {URL} from 'url';
 
-var common = require('./common'/*.js*/);
-var semver = require('semver');
-var stream = require('stream');
+import {TraceSpan} from '../../src/trace';
+import * as trace from '../trace';
+import {ASSERT_SPAN_TIME_TOLERANCE_MS, assertSpanDuration, DEFAULT_SPAN_DURATION} from '../utils';
+import {Express4} from '../web-frameworks/express';
 
-require('../../..').start({
-  projectId: '0',
-  samplingRate: 0
-});
+// This type describes (http|https).(get|request).
+type HttpRequest =
+    (options: string|httpModule.RequestOptions|httpsModule.RequestOptions|URL,
+     callback?: (res: httpModule.IncomingMessage) => void) =>
+        httpModule.ClientRequest;
 
-var assert = require('assert');
-
-describe('test-trace-http', function() {
-  var http;
-  var server;
-  before(function() {
-    http = require('http');
+/**
+ * A class that represents a convenience object that allows us to await
+ * http requests. This is done in lieu of promisifying http.request/http.get,
+ * because they already have a meaningful return value.
+ */
+class WaitForResponse {
+  private resolve!: (value: string) => void;
+  private reject!: (err: Error) => void;
+  // A Promise that is resolved when the request function to which
+  // this.handleResponse is passed has received its full response, or
+  // this.handleDone has been called.
+  done: Promise<string>;
+  // A callback to be passed to http.request or http.get, so that when response
+  // data has been fully consumed, this.done will be resolved.
+  handleResponse = (res: httpModule.IncomingMessage) => {
+    let data = '';
+    res.on('data', d => data += d);
+    res.on('error', this.reject);
+    res.on('end', () => this.resolve(data));
+  };
 
-    server = http.Server(function(req, res) {
-      setTimeout(function() {
-        res.writeHead(200);
-        res.end(common.serverRes);
-      }, common.serverWait);
+  constructor() {
+    this.done = new Promise((resolve, reject) => {
+      this.resolve = resolve;
+      this.reject = reject;
     });
-  });
+  }
 
-  afterEach(function() {
-    common.cleanTraces();
-    server.close();
-  });
+  /**
+   * Resolves this.done. This should be used when passing handleResponse
+   * to a function isn't applicable.
+   */
+  handleDone() {
+    this.resolve('');
+  }
+}
+
+const wait = (ms: number) => new Promise(resolve => setTimeout(resolve, ms));
 
-  it('should patch the necessary functions', function() {
-    assert.strictEqual(http.request.__wrapped, true);
-    if (semver.satisfies(process.version, '>=8.0.0')) {
-      assert.strictEqual(http.get.__wrapped, true, 'should patch get');
-    } else {
-      assert.strictEqual(http.get.__wrapped, undefined, 'should not patch get');
+// Server abstraction class definitions. These are borrowed from web framework
+// tests -- which are useful because they already expose a Promise API.
+const servers = {
+  http: Express4, https: class Express4Secure extends Express4 {
```

Comment:
Formatting is weird. Can you append a comma at the end of line 105 and reformat?

Reply:
Comma doesn't help, and for the record reversing the order of `http` and `https` results in even more weirdness.

I've moved `class Express4Secure` out be declared immediately before, that helps.